src/core/column/cummax.h

```diff
@@ -0,0 +1,93 @@
+//------------------------------------------------------------------------------
+// Copyright 2022 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_CUMMAX_h
+#define dt_COLUMN_CUMMAX_h
+#include "column/virtual.h"
+#include "parallel/api.h"
+#include "stype.h"
+
+namespace dt {
+
+
+template <typename T>
+class Cummax_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Column col_;
+    Groupby gby_;
+
+  public:
+    Cummax_ColumnImpl(Column&& col, const Groupby& gby)
+      : Virtual_ColumnImpl(col.nrows(), col.stype()),
+        col_(std::move(col)),
+        gby_(gby)
+    {
+      xassert(col_.can_be_read_as<T>());
+    }
+
+
+    void materialize(Column& col_out, bool) override {
+      Column col = Column::new_data_column(col_.nrows(), col_.stype());
+      auto data = static_cast<T*>(col.get_data_editable());
+
+      auto offsets = gby_.offsets_r();
+      dt::parallel_for_dynamic(
+        gby_.size(),
+        [&](size_t gi) {
+          size_t i1 = size_t(offsets[gi]);
+          size_t i2 = size_t(offsets[gi + 1]);
+
+          T val;
+          bool is_valid = col_.get_element(i1, &val);
+          data[i1] = is_valid? val : 0;
+
+          for (size_t i = i1 + 1; i < i2; ++i) {
+            is_valid = col_.get_element(i, &val);
+            val = is_valid? val : 0;
+            data[i] = data[i - 1] > val ? data[i - 1] : val;
```

Comment:
Also here, if the element is not valid we should check the previous maximum validity:
- if the previous maximum is not valid, then set the current maximum to be `GETNA<T>()`;
- if the previous maximum is valid, then set the current maximum to be equal to the previous maximum.